tesseract_motion_planners/include/tesseract_motion_planners/simple/profile/simple_planner_utils.h

```diff
@@ -0,0 +1,130 @@
+/**
+ * @file simple_planner_utils.h
+ * @brief Provides interpolation utils for simple planner profiles
+ *
+ * @author Levi Armstrong
+ * @date January 12, 2021
+ * @version TODO
+ * @bug No known bugs
+ *
+ * @copyright Copyright (c) 2021, Southwest Research Institute
+ *
+ * @par License
+ * Software License Agreement (Apache License)
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TESSERACT_MOTION_PLANNERS_SIMPLE_PROFILE_SIMPLE_PLANNER_UTILS_H
+#define TESSERACT_MOTION_PLANNERS_SIMPLE_PROFILE_SIMPLE_PLANNER_UTILS_H
+
+#include <tesseract_common/macros.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_PUSH
+#include <vector>
+#include <memory>
+#include <array>
+TESSERACT_COMMON_IGNORE_WARNINGS_POP
+
+#include <tesseract_command_language/command_language.h>
+#include <tesseract_motion_planners/core/types.h>
+
+namespace tesseract_planning
+{
+/** @brief The Instruction Information struct */
+struct InstructionInfo
+{
+  InstructionInfo(const PlanInstruction& plan_instruction,
+                  const PlannerRequest& request,
+                  const ManipulatorInfo& manip_info);
+
+  const PlanInstruction& instruction;
+  tesseract_kinematics::ForwardKinematics::Ptr fwd_kin;
+  tesseract_kinematics::InverseKinematics::Ptr inv_kin;
+  Eigen::Isometry3d world_to_base;
+  Eigen::Isometry3d working_frame;
+  Eigen::Isometry3d tcp;
+  bool has_cartesian_waypoint{ false };
+
+  /**
+   * @brief Calculate the cartesian pose in world coordinate given the joint solution
+   * @param jp The joint solution to calculate the world pose
+   * @return The world pose give the joint solution
+   */
+  Eigen::Isometry3d calcCartesianWorldPose(const Eigen::VectorXd& jp) const;
+
+  /**
+   * @brief Covert cartesian pose in world to the local robot base frame
+   * @return Cartesian pose in the local robot base frame
+   */
+  Eigen::Isometry3d calcCartesianLocalPose(const Eigen::Isometry3d& world) const;
+
+  /**
+   * @brief Extract the cartesian pose in the world
+   * @details If the instruction does not have a cartesian waypoint this throws an exception
+   * @return Cartesian pose in the world
+   */
+  Eigen::Isometry3d extractCartesianWorldPose() const;
+
+  /**
+   * @brief Extract the joint positiion from the instruction waypoint
+   * @details If the instruction does not have a joint/state waypoint this throws an exception
+   * @return Joint Position
+   */
+  const Eigen::VectorXd& extractJointPosition() const;
+};
```

Comment:
This class technically works, but it seems like a convenient hack for helper functions that really belong in the `CartesianWaypoint` and `JointWaypoint` classes. Classes holding a const reference to some other object also makes me a little suspicious since it can't know about or control when that object goes out of scope. I don't know that you necessarily need to change this now, but I think the planners and other code that leverages waypoint definitions would benefit from these functions being a part of the waypoint implementations

Reply:
This was intentional because the goal of the Command Language was to keep it independent of the other components. Just contain the minimal amount of information need to describe a motion request and not be dependent on other components of Tesseract. This purely meant to only be used internally and and is not intended to be used, if you want I can set the visibility to hidden so it is not in the public API?

Reply:
I don't think changing the visibility of this class really solves the issue. I'm more concerned that other developers might accidentally use it improperly. I would also argue that the couple of functions about Cartesian local and world frames are critical for understanding/working with Cartesian waypoints. The waypoint itself doesn't need to hold on to IK/FK information, but I think it should be able to report it's location in either of those coordinate systems provided a map of transforms and an IK/FK solver.

We don't have to resolve this now, but I just wanted to bring this up as a point of discussion. I can open an issue or discussion to track this further if you want